Format drives in parallel

cholcombe973 opened the pull request:
This takes all the drives specified in juju storage or
manually and runs them through a series of checks to see
if they are a block device, if they are initialized, etc.
After all checks are done in sequential order they are
formatted in parallel.

cholcombe973 committed:
Format drives in parallel

This takes all the drives specified in juju storage or
manually and runs them through a series of checks to see
if they are a block device, if they are initialized, etc.
After all checks are done in sequential order they are
formatted in parallel.

cholcombe973 commented:
@ChrisMacNaughton maybe you can come up with some suggestions to tighten this up even more?

cholcombe973 committed:
Only use initialized devices when setting up the volume

cholcombe973 commented:
This is working now in testing.

cholcombe973 committed:
Skip invalid block devices

cholcombe973 committed:
Make the tasks more parallel

cholcombe973 committed:
Use crossbeam scoped threads to check for rayon problems

cholcombe973 committed:
Use crossbeam instead of rayon.  Seems to actually be parallel

ChrisMacNaughton commented on src/hooks/server_changed.rs:
```diff
@@ -1,5 +1,7 @@
+extern crate crossbeam;
 extern crate gluster;
 extern crate juju;
+//extern crate rayon;
```
lets take out the commented out bits

cholcombe973 replied:
Done

ChrisMacNaughton commented on src/hooks/server_changed.rs:
```diff
@@ -296,77 +307,113 @@ fn expand_volume(peers: Vec<Peer>, volume_info: Option<Volume>) -> Result<i32, S
+        log!(format!("Checking if {:?} is a block device", &device_path));
+        let is_block_device = block::is_block_device(&device_path).unwrap_or(false);
+        if !is_block_device {
+            log!("Skipping invalid block device");
```
Add the name of the block device to the message?

cholcombe973 replied:
Done

ChrisMacNaughton commented on src/hooks/server_changed.rs:
```diff
@@ -296,77 +307,113 @@ fn expand_volume(peers: Vec<Peer>, volume_info: Option<Volume>) -> Result<i32, S
+    log!(format!("storage devices: {:?}", brick_devices));
+
+    // Format all drives in parallel
+    crossbeam::scope(|scope| for device in &mut brick_devices {
```
Rather than pass in a `&mut brick_devices`, can crossbeam give you something equivalent to a `map` so that you could just have the result of this block be a list of initialized bricks?

cholcombe973 replied:
No the api can't do that :(

ChrisMacNaughton reviewed (commented):
A few inline comments

cholcombe973 committed:
Add logging to show which devices are being skipped

cholcombe973 committed:
Fix typo

cholcombe973 committed:
Try using futures instead

cholcombe973 committed:
Futures did not work.  Back to just spawning children

cholcombe973 commented:
@ChrisMacNaughton so yeah.. I dumped futures and tokio.  I'm back to just spawning children in the background.  This works.  I'm going to test it on a bigger box that has multiple drives just to confirm.

cholcombe973 commented:
Doug made an excellent suggestion and I snuck a ps aux command in right after I spawned all the mkfs commands.  It is indeed parallel https://gist.githubusercontent.com/cholcombe973/b26f1adb5aaa18b361bb3bfdb4e04625/raw/ed181c01edb172886f2c0df2c8a492eeac4cd997/gistfile1.txt

ChrisMacNaughton commented on src/hooks/brick_attached.rs:
```diff
@@ -4,10 +4,13 @@ use std::path::PathBuf;
-    initialize_storage(&brick_path)?;
-    Ok(())
+    //initialize_storage(&brick_path)?;
+    //Ok(())
```
Looks like these comments are not needed

cholcombe973 replied:
Yeah I decided to see what would happen if I skipped the brick_attached hook and did the juju storage drives in parallel.  It worked fine so I'm just going to remove that hook.

ChrisMacNaughton commented on src/hooks/config_changed.rs:
```diff
@@ -29,50 +29,20 @@ fn check_for_new_devices() -> Result<(), String> {
-            }
-        }
+        log!("Checking for ephemeral unmount");
+        ephemeral_unmount()?;
```
Do we want to fail the hook when `ephemeral_unmount` errors?

cholcombe973 replied:
That's a good question.  I can't decide.  At one point it I had it failing the hook, then I had it skipped and now I'm back to failing it.  I think the problem is I don't have an association of mountpoint to device yet.  That's going to change soon because I'm working on a fstab parser/modifier.

ChrisMacNaughton commented on src/hooks/server_changed.rs:
```diff
@@ -296,77 +295,74 @@ fn expand_volume(peers: Vec<Peer>, volume_info: Option<Volume>) -> Result<i32, S
-        let mount_path = format!("/mnt/{}", brick_filename.to_string_lossy());
+    // TODO: Should this fail the hook or just keep going?
+    log!("Checking for ephemeral unmount");
+    ephemeral_unmount().map_err(|e| Status::InvalidConfig(e))?;
```
Do we really need to `ephemeral_unmount` both here and in the above hook?

cholcombe973 replied:
Good point.  I'll just stick with server_changed to use it for now.

ChrisMacNaughton commented on src/hooks/server_changed.rs:
```diff
@@ -296,77 +295,74 @@ fn expand_volume(peers: Vec<Peer>, volume_info: Option<Volume>) -> Result<i32, S
-        log!(format!("brick_path: {:?}", brick_path));
+    // Get the juju storage block devices
+    let juju_config_brick_devices = block::get_juju_bricks().map_err(|e| Status::InvalidConfig(e))?;
+    brick_devices.extend(juju_config_brick_devices);
```
This is the second place I've seen this code to collect all of the block devices, would a shared helper function help?

cholcombe973 replied:
Yeah both the config_changed hook and server_changed do almost exactly the same thing.  I can certainly make a helper to clean up some of these calls

ChrisMacNaughton reviewed (approved):
Looks fine, just a few discussion points

cholcombe973 committed:
Add/Remove entries from the fstab as needed

cholcombe973 committed:
Fix config_changed

After a deploy it became obvious that config_changed
is needed on the subordinate machines or they won't
format any of their drives.  This also adds fstab
entries for any drives it formats and mounts so
they can survive a reboot.

cholcombe973 commented:
@ChrisMacNaughton so i ran into a problem where only the leader executes server_changed and all the subordinates execute config_changed.  I think that's why I had that code duplicated

cholcombe973 commented:
I'm going to land this sucker and then put up a new PR for fixing some little things.  This PR is getting rather long
